src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs

```diff
@@ -246,115 +246,159 @@ private XmlQualifiedName GetPrimitiveTypeName(Type type)
         [RequiresUnreferencedCode(XmlSerializer.TrimSerializationWarning)]
         protected void WriteTypedPrimitive(string? name, string? ns, object o, bool xsiType)
         {
-            string? value;
-            string type;
             string typeNs = XmlSchema.Namespace;
-            bool writeRaw = true;
-            bool writeDirect = false;
+            string type;
+
+            void WriteFunc<T>(T value)
+            {
+                if (name == null)
+                    _w.WriteStartElement(type, typeNs);
+                else
+                    _w.WriteStartElement(name, ns);
+
+                if (xsiType) WriteXsiType(type, typeNs);
+
+                if (value == null)
+                {
+                    _w.WriteAttributeString("nil", XmlSchema.InstanceNamespace, "true");
+                }
+                else
+                {
+                    if (typeof(T) == typeof(string))
+                        _w.WriteString((string?)(object?)value);
+                    else if (typeof(T) == typeof(int))
+                        _w.WriteValue((int)(object)value);
+                    else if (typeof(T) == typeof(bool))
+                        _w.WriteValue((bool)(object)value);
+                    else if (typeof(T) == typeof(short))
+                        _w.WriteValue((short)(object)value);
+                    else if (typeof(T) == typeof(long))
+                        _w.WriteValue((long)(object)value);
+                    else if (typeof(T) == typeof(float))
+                        _w.WriteValue((float)(object)value);
+                    else if (typeof(T) == typeof(double))
+                        _w.WriteValue((double)(object)value);
+                    else if (typeof(T) == typeof(decimal))
+                        _w.WriteValue((decimal)(object)value);
+                    else if (typeof(T) == typeof(byte))
+                        _w.WriteValue((byte)(object)value);
+                    else if (typeof(T) == typeof(sbyte))
+                        _w.WriteValue((sbyte)(object)value);
+                    else if (typeof(T) == typeof(ushort))
+                        _w.WriteValue((ushort)(object)value);
+                    else if (typeof(T) == typeof(uint))
+                        _w.WriteValue((uint)(object)value);
+                    //else if (typeof(T) == typeof(ulong))
```

Comment:
If we want to support these also - we need to extend XmlWriter.WriteValue methods with ulong and guid